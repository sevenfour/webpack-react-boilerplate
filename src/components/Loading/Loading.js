import React from 'react';
import PropTypes from 'prop-types';

import clsx from 'clsx';

import { Circle } from '../icons';

import styles from './Loading.module.css';
import stylesUtilities from 'styles/Utilities.module.css';

const Loading = props => {

  const {
    className,
    overlay,
    ...rest
  } = props;

  const loader = (
    <div
      className={clsx(styles.spinner, className)}
      role="progressbar"
      {...rest}
    >
      <p className={stylesUtilities.visuallyHidden}>
        Content is loading
      </p>
      <Circle
        className={styles.circleIcon}
        color="transparent"
      />
    </div>
  );

  return overlay
    ? <div className="overlay">{loader}</div>
    : loader;
};

Loading.propTypes = {
  className: PropTypes.string,
  overlay: PropTypes.bool
};

export default Loading;
