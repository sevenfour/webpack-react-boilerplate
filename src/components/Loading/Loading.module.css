
/*
 * The animation principle is taken from Material-UI
 * https://github.com/mui-org/material-ui/blob/next/packages/material-ui/src/CircularProgress/CircularProgress.js
*/

.spinner {
  display: inline-block;
  width: 40px;
  height: 40px;
  animation: rotate 1.4s infinite linear;
  color: var(--grass-green);
}

.circleIcon {
  width: inherit;
  height: inherit;
}

.circleIcon > circle {
  stroke: currentColor;
  stroke-dasharray: 80px, 200px;
  stroke-dashoffset: 0;
  animation: dash 1.4s infinite ease-in-out;
}

.overlay {
  display: grid;
  opacity: 0.75;
  z-index: 9999;
  place-items: center;
  background-color: var(--grey-alpha-10);
  width: 100%;
  height: 100%;
}

@keyframes rotate {

  0% {
    transform-origin: 50% 50%;
  }

  100% {
    transform: rotate(360deg);
  }
}

@keyframes dash {

  0% {
    stroke-dasharray: 1px, 200px;
    stroke-dashoffset: 0;
  }

  50% {
    stroke-dasharray: 100px, 200px;
    stroke-dashoffset: -15px;
  }

  100% {
    stroke-dasharray: 100px, 200px;
    stroke-dashoffset: -125px;
  }
}

@media ( prefers-reduced-motion ) {

  .spinner {
    animation: none;
  }

  .circleIcon > circle {
    animation: none;
  }
}
