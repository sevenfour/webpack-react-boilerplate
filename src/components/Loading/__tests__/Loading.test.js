import React from 'react';
import { shallow } from 'enzyme';

import Loading from '..';

describe('<Loading /> component tests', () => {

  it('renders', () => {
    const loader = shallow(
      <Loading />
    );

    expect(loader).toMatchSnapshot();
  });

});
