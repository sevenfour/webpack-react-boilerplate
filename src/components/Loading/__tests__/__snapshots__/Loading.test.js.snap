// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading /> component tests renders 1`] = `
<div
  className="spinner"
  role="progressbar"
>
  <p
    className="visuallyHidden"
  >
    Content is loading
  </p>
  <Circle
    className="circleIcon"
    color="transparent"
  />
</div>
`;
